### PR TITLE
Feature/fdp 46 mqtt ssl

### PR DESCRIPTION
**Pull request by smvdheijden**

### CODE REVIEW

**Comment by smvdheijden** [comment]
Retest this please.

**Comment by smvdheijden** [comment]
Retest this please.

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/SubscriptionService.java [hunk 1]
```diff
@@ -152,4 +146,17 @@ public void onReceive(final MqttClientAdapter mqttClientAdapter, final byte[] pa
             .build();
     this.outboundOsgpCoreResponseMessageSender.send(responseMessage);
   }
+
+  private int getDefaultPort() {
+    return Integer.valueOf(
+        this.mqttClientProperties.getProperty(MqttConstants.DEFAULT_PORT_PROPERTY_NAME));
```
Comment (bvdzwet) [inline osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/SubscriptionService.java:152]: `Integer.parseInt` would fit the `int` result a little better.
Looks like a `NullPointerException` could occur is the `mqttClientProperties` did not have the specific property obtained here. These kind of things may be caught earlier with a more specific input object than `Properties` that could already do some validation on construction.

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/resources/osgp-adapter-protocol-mqtt.properties [hunk 2]
```diff
@@ -31,9 +31,17 @@ flyway.init.on.migrate=true
...
-mqtt.broker.defaultPort=8883
-mqtt.broker.defaultTopics=+/measurement,+/congestion
-mqtt.broker.defaultQos=AT_LEAST_ONCE
+mqtt.default.host=localhost
+mqtt.default.port=8883
+mqtt.default.qos=AT_LEAST_ONCE
+mqtt.default.topics=+/measurement,+/congestion
+
+mqtt.ssl.enabled=true
+#mqtt.ssl.client.truststore.location=/etc/osp/keys/mqtt/mqtt_client_truststore.jks
```
Comment (bvdzwet) [inline osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/resources/osgp-adapter-protocol-mqtt.properties:40]: Any reason to keep alternatives commented out. Looks clear enough if this were removed.
If it is intended as documentation, a more explicit comment about allowed values like: "may contain a file name on the classpath, or the full path to the desired file" may be a better fit. A switch from String to Resource would also work, in which case the value already gives a hint that it describes some form of resource (starting with 'classpath:' for instance.)

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/test/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/SubscriptionServiceTest.java [hunk 3]
```diff
@@ -53,21 +55,33 @@
...
+
+    lenient()
+        .when(this.mqttClientProperties.getProperty(MqttConstants.DEFAULT_PORT_PROPERTY_NAME))
+        .thenReturn(String.valueOf(DEFAULT_PORT));
+    lenient()
+        .when(this.mqttClientProperties.getProperty(MqttConstants.DEFAULT_QOS_PROPERTY_NAME))
+        .thenReturn(DEFAULT_QOS.name());
+    lenient()
+        .when(this.mqttClientProperties.getProperty(MqttConstants.DEFAULT_TOPICS_PROPERTY_NAME))
+        .thenReturn(DEFAULT_TOPICS);
```
Comment (bvdzwet) [inline osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/test/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/SubscriptionServiceTest.java:80]: As long as `Properties` are used, as far as the test results go, this works alright. But in a case like this wouldn't just setting up the actual properties  lead to simpler code than mocking the expected return values?

File: osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/java/org/opensmartgridplatform/simulator/protocol/mqtt/Client.java [hunk 4]
```diff
@@ -27,30 +37,75 @@
...
+      LOG.info("Load default truststore from classpath: {}", trustStorePath);
+      trustStore = loadTrustStoreFromClassPath(trustStoreType, trustStorePath, trustStorePW);
+    }
+
+    final TrustManagerFactory tmf =
+        TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+    tmf.init(trustStore);
+
+    return tmf;
   }
```
Comment (bvdzwet) [inline osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/java/org/opensmartgridplatform/simulator/protocol/mqtt/Client.java:109]: This looks almost identical to the code in `MqttClientAdapter` (lines 96 to 130) in osgp-protocol-adapter-mqtt.
Some form of more specific keystore configuration object, or maybe even MQTT specific could probably remove the duplication and simplify the code where it is used.
The code for alternatives for loading the keystore files (140-167) would fall under this same comment I'd guess.

File: osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/java/org/opensmartgridplatform/simulator/protocol/mqtt/SimulatorConfig.java [hunk 5]
```diff
@@ -9,25 +9,38 @@
...
 
   private static final Logger LOG = LoggerFactory.getLogger(SimulatorConfig.class);
 
+  @Value("${mqtt.simulator.ssl.enabled:false}")
+  private boolean sslEnabled;
+
+  @Bean
+  public Boolean isSslEnabled() {
+    return this.sslEnabled;
+  }
```
Comment (bvdzwet) [inline osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/java/org/opensmartgridplatform/simulator/protocol/mqtt/SimulatorConfig.java:32]: Is this used. Looking at `SimulatorBrokerConfig` and `SimulatorClientConfig` where the `mqttBrokerProperties` and `mqttClientProperties` come from, they each have their own `boolean sslEnabled`.
Not sure why there is a default value in the broker config that is not explicitly set in the client config (though with the value being `false` the effect appears the same).

File: osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/resources/osgp-simulator-protocol-mqtt.properties [hunk 6]
```diff
@@ -1,3 +1,21 @@
...
+mqtt.simulator.startClient=true
+
+mqtt.simulator.broker.host=localhost
+mqtt.simulator.broker.port=1883
+
+mqtt.simulator.ssl.enabled=true
+
+mqtt.simulator.broker.ssl.port=8883
+mqtt.simulator.broker.ssl.provider=JDK
+#mqtt.simulator.broker.ssl.keystore.location=/etc/osp/keys/mqtt/mqtt_broker_keystore.jks
```
Comment (bvdzwet) [inline osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/resources/osgp-simulator-protocol-mqtt.properties:12]: Probably cleaner not to leave in commented out alternatives like here and in line 17.

**Review by bvdzwet — CHANGES_REQUESTED** [review]

File: integration-tests/cucumber-tests-platform-distributionautomation/src/test/java/org/opensmartgridplatform/cucumber/platform/distributionautomation/glue/steps/MqttDeviceSteps.java [hunk 7]
```diff
@@ -72,7 +74,29 @@ public void theDeviceSendsAMeasurementReport(
...
     final Message[] messages = {message};
     spec.setMessages(messages);
-    final SimulatorSpecPublishingClient publishingClient = new SimulatorSpecPublishingClient(spec);
+
+    final Properties properties = new Properties();
+
+    properties.put(MqttConstants.DEFAULT_HOST_PROPERTY_NAME, host);
+    properties.put(MqttConstants.DEFAULT_PORT_PROPERTY_NAME, port);
+    properties.put(
+        MqttConstants.DEFAULT_QOS_PROPERTY_NAME, PlatformDistributionAutomationDefaults.MQTT_QOS);
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply integration-tests/cucumber-tests-platform-distributionautomation/src/test/java/org/opensmartgridplatform/cucumber/platform/distributionautomation/glue/steps/MqttDeviceSteps.java]: Refactored.

File: integration-tests/cucumber-tests-platform-distributionautomation/src/test/java/org/opensmartgridplatform/cucumber/platform/distributionautomation/glue/steps/MqttDeviceSteps.java [hunk 8]
```diff
@@ -72,7 +74,29 @@ public void theDeviceSendsAMeasurementReport(
     final Message message = new Message(topic, payload, 10000);
     final Message[] messages = {message};
     spec.setMessages(messages);
-    final SimulatorSpecPublishingClient publishingClient = new SimulatorSpecPublishingClient(spec);
+
+    final Properties properties = new Properties();
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply integration-tests/cucumber-tests-platform-distributionautomation/src/test/java/org/opensmartgridplatform/cucumber/platform/distributionautomation/glue/steps/MqttDeviceSteps.java]: Refactored.

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java [hunk 9]
```diff
@@ -8,37 +8,56 @@
...
+      final Properties mqttClientProperties,
       final MqttClientEventHandler mqttClientEventHandler) {
     this.device = device;
     this.messageMetadata = messageMetadata;
+    this.mqttClientProperties = mqttClientProperties;
     this.mqttClientEventHandler = mqttClientEventHandler;
   }
 
-  public void connect() {
+  public void connect() throws GeneralSecurityException {
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java]: Refactored.

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java [hunk 10]
```diff
@@ -73,4 +92,70 @@ public MqttDevice getDevice() {
...
+  }
+
+  private static TrustManagerFactory getTruststoreFactory(final Properties mqttProperties)
+      throws GeneralSecurityException {
+
+    final String trustStoreType =
+        mqttProperties.getProperty(MqttConstants.SSL_TRUSTSTORE_TYPE_PROPERTY_NAME);
+    final String trustStorePath =
+        mqttProperties.getProperty(MqttConstants.SSL_TRUSTSTORE_PATH_PROPERTY_NAME);
+    final String trustStorePW =
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java]: Refactored.

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java [hunk 11]
```diff
@@ -73,4 +92,70 @@ public MqttDevice getDevice() {
...
+        mqttProperties.getProperty(MqttConstants.SSL_TRUSTSTORE_PASSWORD_PROPERTY_NAME);
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java]: Refactored.

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java [hunk 12]
```diff
@@ -73,4 +92,70 @@ public MqttDevice getDevice() {
...
+
+    KeyStore trustStore;
+    if (Files.exists(Paths.get(trustStorePath))) {
+      LOGGER.info("Load external truststore from path: {}", trustStorePath);
+      trustStore = loadTrustStoreFromExternalPath(trustStoreType, trustStorePath, trustStorePW);
+    } else {
+      LOGGER.info("Load default truststore from classpath: {}", trustStorePath);
+      trustStore = loadTrustStoreFromClassPath(trustStoreType, trustStorePath, trustStorePW);
+    }
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java]: Refactored.

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java [hunk 13]
```diff
@@ -73,4 +92,70 @@ public MqttDevice getDevice() {
...
+
+  private static KeyStore loadTrustStoreFromExternalPath(
+      final String trustStoreType, final String trustStorePath, final String trustStorePW)
+      throws GeneralSecurityException {
+
+    try (InputStream in = new FileInputStream(trustStorePath)) {
+      final KeyStore trustStore = KeyStore.getInstance(trustStoreType);
+      trustStore.load(in, trustStorePW.toCharArray());
+      return trustStore;
+    } catch (final Exception e) {
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/MqttClientAdapter.java]: Refactored.

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/resources/osgp-adapter-protocol-mqtt.properties (hunk 2 above)
Comment (smvdheijden) [reply osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/main/resources/osgp-adapter-protocol-mqtt.properties]: Updated comments.

File: osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/test/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/SubscriptionServiceTest.java (hunk 3 above)
Comment (smvdheijden) [reply osgp/protocol-adapter-mqtt/osgp-protocol-adapter-mqtt/src/test/java/org/opensmartgridplatform/adapter/protocol/mqtt/application/services/SubscriptionServiceTest.java]: Refactored.

File: osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/resources/osgp-simulator-protocol-mqtt.properties (hunk 6 above)
Comment (smvdheijden) [reply osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/resources/osgp-simulator-protocol-mqtt.properties:13]: Updated comments

File: osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/java/org/opensmartgridplatform/simulator/protocol/mqtt/Client.java (hunk 4 above)
Comment (smvdheijden) [reply osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/java/org/opensmartgridplatform/simulator/protocol/mqtt/Client.java]: Refactored

**Comment by smvdheijden** [comment]
Retest this please.

**Review by bvdzwet — APPROVED** [review]

**Pull request merged**